apps/gateway/lib/gateway/controllers/curse_of_mirra/character_controller.ex

```diff
@@ -21,4 +21,35 @@ defmodule Gateway.Controllers.CurseOfMirra.CharacterController do
       send_resp(conn, 200, Jason.encode!(%{character_name: params["character_name"], skin_name: params["skin_name"]}))
     end
   end
+
+  def level_up(conn, params) do
+    with {:ok, unit} <- Units.get_unit_by_character_name(params["character_name"], params["user_id"]),
+         {:ok, %{unit: leveled_up_unit, user_currency: user_currencies}} <- Units.level_up(params["user_id"], unit.id) do
+      new_balances =
+        user_currencies
+        |> Enum.map(fn user_currency ->
+          %{amount: user_currency.amount, currency: %{name: user_currency.currency.name}}
+        end)
+
+      send_resp(
+        conn,
+        200,
+        Jason.encode!(%{
+          character_name: params["character_name"],
+          new_level: leveled_up_unit.level,
+          new_currency_balance: new_balances
+        })
+      )
+    else
+      {:error, :no_more_levels} ->
+        send_resp(conn, 400, "Cannot further level up")
+
+      {:error, :cant_afford} ->
+        send_resp(conn, 400, "Cannot afford levelling up")
+    end
```

Comment:
Can we use the fallback controller instead of this? I don't know when those happen.
Also, we already have there an error clause for "cant_afford" cases.
I think a more understandable solution should be to check if the user has the enough currency to do this request, if that's ok, then it can level up.
That's how is being done everywhere, and that's why we use `with` + FallbackController, so we can split the requests in sequences of condition-checks then trigger the proper error accordingly.

Reply:
done! By this do you mean avoid doing the request from the frontend since it already knows you don't have enough money?